sshd/zenith/sshd/backends/crd.py

```diff
@@ -0,0 +1,182 @@
+                "metadata": {
+                    # Generate a name based on the subdomain
+                    # We will use this as the tunnel ID
+                    "generateName": f"{subdomain}-",
+                    "ownerReferences": [
+                        {
+                            "apiVersion": endpoints["apiVersion"],
+                            "kind": endpoints["kind"],
+                            "name": endpoints["metadata"]["name"],
+                            "uid": endpoints["metadata"]["uid"],
+                            "blockOwnerDeletion": True,
+                        },
+                    ],
+                },
+                "spec": {
+                    "renewedAt": isotime(),
+                    "ttl": ttl,
+                    "reapAfter": reap_after,
+                }
+            }
+        )
+
+        # The tunnel ID is the lease name with the subdomain prefix removed
+        tunnel_id = lease["metadata"]["name"].removeprefix(f"{subdomain}-")
+
+        # Update the endpoints resource with the endpoint definition
+        # The endpoints resource should already exist
+        ekendpoints.patch(
+            subdomain,
+            {
+                "spec": {
+                    "endpoints": {
+                        tunnel_id: {
+                            "address": host,
+                            "port": port,
+                            # The initial status is critical, until the first heartbeat
+                            "status": base.TunnelStatus.CRITICAL.value,
+                            "config": config_dict,
+                        },
+                    },
+                },
+            }
+        )
+    
+        return tunnel_id
+
+    def tunnel_heartbeat(self, subdomain: str, id: str, status: base.TunnelStatus):
+        # Renew the lease
+        self.ekclient.api(self.api_version).resource("leases").patch(
+            f"{subdomain}-{id}",
+            {
+                "spec": {
+                    "renewedAt": isotime(),
+                },
+            }
+        )
+        # Update the endpoint status
+        self.ekclient.api(self.api_version).resource("endpoints").patch(
+            subdomain,
+            {
+                "spec": {
+                    "endpoints": {
+                        id: {
+                            "status": status.value,
+                        },
+                    },
+                },
+            }
+        )
+
+    def tunnel_terminate(self, subdomain: str, id: str):
+        # Remove the tunnel from the endpoints object
+        self.ekclient.api(self.api_version).resource("endpoints").json_patch(
+            subdomain,
+            [
+                {
+                    "op": "remove",
+                    "path": f"/spec/endpoints/{id}",
+                },
+            ]
+        )
+        # Delete the lease
+        self.ekclient.api(self.api_version).resource("leases").delete(f"{subdomain}-{id}")
```

Comment:
Feels like this needs a nice log message?

Reply:
The caller does log this. easykube will also log all the API requests on the server side (not sent to the client).